### DISCUSSION TITLE
Selecting subset of edge indices

### DISCUSSION THREAD

**Discussion by ahmedramly** [pyg-team/pytorch_geometric#3774] [Q&A] [answered]
I have a question about selecting a subset of edge indices. I have the following situation, I have `14 Nodes` and the experiments include trying different combinations of edges and edge attributes so I want to do the following:
- Have the edge indices of a full graph which is of size `torch.Size([2, 196])`
- Have the edge attributes of a full graph which is of size `torch.Size([196, N])` where `N` is the number of edge features
- Using a subset of edge indices of size `torch.Size([2, m])` to define the graph and the edge attributes where `m` is the number of the edges in the subset.

It is a naive question but I am kinda stuck with it for a while. Thanks in advance

**Comment by wsad1** [comment, marked as answer]
One way to do this is as follows.

1. `torch.cartesian_prod(torch.arange(196), torch.arange(196)).T` create a fully connected graph.
2. use [dropout_adj](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.dropout_adj) with training as true, to sample edges.

**Reply by ahmedramly** [reply]
I think I didn’t clarify my question in a correct way. The goal is not to drop some edges randomly with a probability. The problem is I don’t have the graph structure explicitly so I use different statistical and graph signal processing methods to learn the edges so in code it should be like this 
```
G = Data(x=torch.Size([14, n_feats]), edge_index=torch.Size([2,196]), edge_attr=torch.Size([196,n_attrs]), y=torch.Size([1]))

G.subset_1 = torch.Size([2, 28])
G.subset_2 = torch.Size([2, 34])
```

`subset_1` and ` subset_2` are already included in `edge_index` so the final goal is when training a network to define the edge_index and edge_attr in the `forward` As something like this

```
model(x = x, edge_index=subset_1, edge_attr=edge_attr[G.subset_1])
```
But it doesn’t work like this , I can’t use the `subset_1` edge index as index to extract the corresponding attributes

Did I clarify the problem well?

**Reply by wsad1** [reply]
Initially you have a graph with say 14 nodes and 196 edges (fully connected). Then you predict which edges actually exist , and then run `message passing` using only the predicted edges. Is this an accurate summary. If so you could approach this 2 ways.
1. The edge classifier could return an edge mask. An edge mask would be a `tensor.bool` of shape `num_edges` (196 in your case). Using this you could do the following `model(x = x, edge_index=edge_index[:,edge_mask], edge_attr=edge_attr[edge_mask])`.  This is a clean and efficient approach
2. Another approach is to modify your current code as follows:
`model(x = x, edge_index=subset_1, edge_attr=edge_attr[torch.all(torch.eq(edge_index, subset_1), dim=0)])`

**Reply by rusty1s** [reply]
To chime in, another approach is to convert `subset_1` into a one-dimensional vector which can be used to query `edge_attr`, e.g.:
```python
idx_1 = subset_1[0] * 14 + subset_1[1]
edge_attr = edge_attr[idx_1]
```
Note that this approach only works in case `edge_index` and `edge_attr` encode a fully-connected graph though :)

**Reply by ahmedramly** [reply]
Thank you all for the help, I tried all of them and most convenient one is the last one by @rusty1s . 
Also I should mention that the second solution by @wsad1 didn’t work with me, it gave me this error
`RuntimeError: The size of tensor a (196) must match the size of tensor b (28) at non-singleton dimension 1`
Thank you for the help again :)
